session/session.go

```diff
@@ -2506,9 +2512,9 @@ func (s *session) isTxnRetryable() bool {
 	}
 
 	// If the session is not InTxn, it is an auto-committed transaction.
-	// The auto-committed transaction could always retry.
+	// The auto-committed transaction could always retry when @@tidb_snapshot is not set.
```

Comment:
Why?

Reply:
`isTxnRetryable` returns whether the txn should retry in optimistic mode. When @@tidb_snapshot is set, it is read only, and no need to retry because it will not cause any errors which should retry a transaction.

In this PR `InitTxnWithStartTS` is removed, and the when a txn want to activate a transaction with a snapshot ts, it will use `Txn` instead of `InitTxnWithStartTS`. This code here also make the behavior same with the old one.

I'll add some comments for it